src/librbd/internal.cc

```diff
@@ -1850,29 +1851,65 @@ int mirror_image_disable_internal(ImageCtx *ictx, bool force,
       }
       assert(m_bl->length() == (size_t)r);
 
-      if (m_bl->is_zero()) {
+      if (!m_sparse_size) {
+	if (m_bl->is_zero()) {
+	  delete m_bl;
+	  m_throttle->end_op(r);
+	  return;
+	}
+
+	Context *ctx = new C_CopyWrite(m_throttle, m_bl);
+	auto comp = io::AioCompletion::create(ctx);
+
+	// coordinate through AIO WQ to ensure lock is acquired if needed
+	m_dest->io_work_queue->aio_write(comp, m_offset, m_bl->length(),
+					 std::move(*m_bl),
+                                         LIBRADOS_OP_FLAG_FADVISE_DONTNEED);
+      } else {
+        m_throttle->end_op(r);
```

Comment:
@mslovy The easiest way to approach it would be to use lambdas (or inherit from C_Gather). If you use lambdas, you would do something like the following:

```
auto *end_op_ctx = new FunctionContext([m_throttle](int r) {
    m_throttle->end_op(r);
  });
auto gather_ctx = new C_Gather(m_dest->cct, end_op_ctx);

for (each chunk to write) {
  ... snip ...
  auto ctx = new C_CopyWrite(m_bl, gather_ctx->new_sub());
  auto comp = io::AioCompletion::create(ctx);
  ... snip ...
}
```

This allows multiple writes to be issued within the same "throttled" op slot, prevents starvation,